// @botonic/react/lib/cjs/contexts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebchatContext = exports.WebviewRequestContext = exports.RequestContext = void 0;
const react_1 = require("react");
const hooks_1 = require("./webchat/hooks");
exports.RequestContext = (0, react_1.createContext)({
    getString: () => '',
    setLocale: () => undefined,
    session: {},
    params: {},
    input: {},
    defaultDelay: 0,
    defaultTyping: 0,
});
exports.WebviewRequestContext = (0, react_1.createContext)({
    closeWebview: () => undefined,
    getString: () => '',
    params: {},
    session: {},
});
exports.WebchatContext = (0, react_1.createContext)({
    addMessage: () => {
        return;
    },
    closeWebview: () => {
        return;
    },
    getThemeProperty: () => {
        return;
    },
    openWebview: () => {
        return;
    },
    resolveCase: () => {
        return;
    },
    resetUnreadMessages: () => {
        return;
    },
    setLastMessageVisible: () => {
        return;
    },
    sendAttachment: () => {
        return;
    },
    sendInput: () => {
        return;
    },
    sendPayload: () => {
        return;
    },
    sendText: () => {
        return;
    },
    theme: {},
    toggleWebchat: () => {
        return;
    },
    updateLatestInput: () => {
        return;
    },
    updateMessage: () => {
        return;
    },
    updateReplies: () => {
        return;
    },
    updateUser: () => {
        return;
    },
    updateWebchatDevSettings: () => {
        return;
    },
    webchatState: hooks_1.webchatInitialState,
});
