// @botonic/react/lib/cjs/webchat/message-list/styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContainerUnreadMessagesBanner = exports.ContainerScrollButton = exports.DefaultIntroImage = exports.ContainerMessage = void 0;
const tslib_1 = require("tslib");
const styled_components_1 = tslib_1.__importDefault(require("styled-components"));
exports.ContainerMessage = styled_components_1.default.div `
  display: flex;
  overflow-x: hidden;
  flex-direction: column;
  flex: none;
  white-space: pre;
  word-wrap: break-word;
`;
exports.DefaultIntroImage = styled_components_1.default.img `
  max-height: 50%;
  width: 100%;
`;
exports.ContainerScrollButton = styled_components_1.default.div `
  position: absolute;
  right: 10px;
  bottom: 65px;

  background-color: #6d6a78;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;

  width: 37px;
  height: 37px;
  border-radius: 50%;
`;
exports.ContainerUnreadMessagesBanner = styled_components_1.default.div `
  display: flex;
  justify-content: center;
  align-items: center;
  gap: 5px;

  border-top: 1px solid #e8e8ea;
  padding: 8px;
  color: #6d6a78;
  font-size: 14px;
  font-weight: 400;
  width: 100%;

  img {
    width: 10px;
  }
`;
