// @botonic/core/lib/cjs/models/legacy-types.js
"use strict";
// TODO: This file contains all the legacy types we had in index.ts. After some refactors, we should be able to get rid of many of them.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Providers = exports.CASE_RESOLUTION = exports.CASE_STATUS = exports.INPUT = exports.PROVIDER = void 0;
var PROVIDER;
(function (PROVIDER) {
    PROVIDER["APPLE"] = "apple";
    PROVIDER["DEV"] = "dev";
    PROVIDER["FACEBOOK"] = "facebook";
    PROVIDER["GENERIC"] = "generic";
    PROVIDER["INSTAGRAM"] = "instagram";
    PROVIDER["INTERCOM"] = "intercom";
    PROVIDER["SMOOCH"] = "smooch";
    PROVIDER["TELEGRAM"] = "telegram";
    PROVIDER["TWITTER"] = "twitter";
    PROVIDER["WEBCHAT"] = "webchat";
    PROVIDER["WECHAT"] = "wechat";
    PROVIDER["WHATSAPP"] = "whatsapp";
})(PROVIDER = exports.PROVIDER || (exports.PROVIDER = {}));
var INPUT;
(function (INPUT) {
    INPUT["TEXT"] = "text";
    INPUT["POSTBACK"] = "postback";
    INPUT["AUDIO"] = "audio";
    INPUT["IMAGE"] = "image";
    INPUT["VIDEO"] = "video";
    INPUT["DOCUMENT"] = "document";
    INPUT["LOCATION"] = "location";
    INPUT["CONTACT"] = "contact";
    INPUT["BUTTON_MESSAGE"] = "buttonmessage";
    INPUT["CAROUSEL"] = "carousel";
    INPUT["CUSTOM"] = "custom";
    INPUT["WEBCHAT_SETTINGS"] = "webchatsettings";
    INPUT["WHATSAPP_TEMPLATE"] = "whatsapptemplate";
    INPUT["RAW"] = "raw";
    INPUT["CHAT_EVENT"] = "chatevent";
    INPUT["WHATSAPP_BUTTON_LIST"] = "whatsapp-button-list";
    INPUT["WHATSAPP_CTA_URL_BUTTON"] = "whatsapp-cta-url-button";
})(INPUT = exports.INPUT || (exports.INPUT = {}));
var CASE_STATUS;
(function (CASE_STATUS) {
    CASE_STATUS["WAITING"] = "status_waiting";
    CASE_STATUS["ATTENDING"] = "status_attending";
    CASE_STATUS["IDLE"] = "status_idle";
    CASE_STATUS["RESOLVED"] = "status_resolved";
})(CASE_STATUS = exports.CASE_STATUS || (exports.CASE_STATUS = {}));
var CASE_RESOLUTION;
(function (CASE_RESOLUTION) {
    CASE_RESOLUTION["OK"] = "result_ok";
    CASE_RESOLUTION["NOK"] = "result_nok";
    CASE_RESOLUTION["NOT_SOLVED"] = "result_not_solved";
    CASE_RESOLUTION["BANNED"] = "result_banned";
})(CASE_RESOLUTION = exports.CASE_RESOLUTION || (exports.CASE_RESOLUTION = {}));
exports.Providers = Object.freeze({
    Messaging: {
        APPLE: 'apple',
        FACEBOOK: 'facebook',
        GENERIC: 'generic',
        IMBEE: 'imbee',
        INSTAGRAM: 'instagram',
        INTERCOM: 'intercom',
        SMOOCH_WEB: 'smooch_web',
        SMOOCH: 'smooch',
        TELEGRAM: 'telegram',
        TWITTER: 'twitter',
        WEBCHAT: 'webchat',
        WECHAT: 'wechat',
        WHATSAPP: 'whatsapp',
    },
});
