// @botonic/react/lib/esm/webchat/actions.js
export var WebchatAction;
(function (WebchatAction) {
    WebchatAction["ADD_MESSAGE"] = "addMessage";
    WebchatAction["ADD_MESSAGE_COMPONENT"] = "addMessageComponent";
    WebchatAction["CLEAR_MESSAGES"] = "clearMessages";
    WebchatAction["DO_RENDER_CUSTOM_COMPONENT"] = "doRenderCustomComponent";
    WebchatAction["RESET_UNREAD_MESSAGES"] = "resetUnreadMessages";
    WebchatAction["SET_CURRENT_ATTACHMENT"] = "setCurrentAttachment";
    WebchatAction["SET_ERROR"] = "setError";
    WebchatAction["SET_LAST_MESSAGE_VISIBLE"] = "setlastMessageVisible";
    WebchatAction["SET_ONLINE"] = "setOnline";
    WebchatAction["TOGGLE_COVER_COMPONENT"] = "toggleCoverComponent";
    WebchatAction["TOGGLE_EMOJI_PICKER"] = "toggleEmojiPicker";
    WebchatAction["TOGGLE_PERSISTENT_MENU"] = "togglePersistentMenu";
    WebchatAction["TOGGLE_WEBCHAT"] = "toggleWebchat";
    WebchatAction["UPDATE_DEV_SETTINGS"] = "updateDevSettings";
    WebchatAction["UPDATE_HANDOFF"] = "updateHandoff";
    WebchatAction["UPDATE_JWT"] = "updateJwt";
    WebchatAction["UPDATE_LAST_MESSAGE_DATE"] = "updateLastMessageDate";
    WebchatAction["UPDATE_LAST_ROUTE_PATH"] = "updateLastRoutePath";
    WebchatAction["UPDATE_LATEST_INPUT"] = "updateLatestInput";
    WebchatAction["UPDATE_MESSAGE"] = "updateMessage";
    WebchatAction["UPDATE_REPLIES"] = "updateReplies";
    WebchatAction["UPDATE_SESSION"] = "updateSession";
    WebchatAction["UPDATE_THEME"] = "updateTheme";
    WebchatAction["UPDATE_TYPING"] = "updateTyping";
    WebchatAction["UPDATE_WEBVIEW"] = "updateWebview";
})(WebchatAction || (WebchatAction = {}));
