// @botonic/core/lib/cjs/data-provider/dynamodb-data-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DynamoDBDataProvider = void 0;
const tslib_1 = require("tslib");
const models_1 = require("../models");
const dynamodb_utils_1 = require("./dynamodb-utils");
class DynamoDBDataProvider {
    constructor(url) {
        try {
            const urlParts = url.split('://')[1].split('.');
            this.tableName = urlParts[0];
            this.region = urlParts[1];
            this.userEventsTable = (0, dynamodb_utils_1.getUserEventsTable)(this.tableName, this.region);
            this.userEntity = (0, dynamodb_utils_1.getUserEntity)(this.userEventsTable);
            this.connectionEventEntity = (0, dynamodb_utils_1.getConnectionEventEntity)(this.userEventsTable);
            this.messageEventEntities = (0, dynamodb_utils_1.getMessageEventEntities)(this.userEventsTable);
        }
        catch (e) {
            console.log({ e });
        }
    }
    // @ts-ignore
    getUsers(limit = 10, offset = 0) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            // TODO: finish to implement with offset
            const result = yield this.userEventsTable.scan({
                filters: { attr: dynamodb_utils_1.SORT_KEY_NAME, beginsWith: dynamodb_utils_1.USER_PREFIX },
                limit,
            });
            return result.Items;
        });
    }
    getUser(id) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const userById = {
                id: id,
                SK: id,
            };
            const result = yield this.userEntity.get(userById);
            if (Object.keys(result).length === 0)
                return undefined;
            return result.Item;
        });
    }
    getUserByWebsocketId(websocketId) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const result = yield this.userEventsTable.query(websocketId, {
                index: dynamodb_utils_1.GLOBAL_SECONDARY_INDEX_NAME,
            });
            if (result.Count === 0)
                return undefined;
            return result.Items[0];
        });
    }
    // @ts-ignore
    getUserByField(field, value) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () { });
    } //TODO: Implement
    saveUser(user) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const putUser = Object.assign(Object.assign({}, user), { id: user.id, userId: user.id });
            yield this.userEntity.put(putUser);
            return user;
        });
    }
    updateUser(user) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const updateUser = Object.assign(Object.assign({}, user), { id: user.id, userId: user.id });
            const res = yield this.userEntity.update(updateUser);
            return user;
        });
    }
    // @ts-ignore
    deleteUser(id) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () { });
    } // TODO: Implement
    // @ts-ignore
    getEvents(limit = 10, offset = 0) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () { });
    } // TODO: Implement
    // @ts-ignore
    getEvent(id) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () { });
    } // TODO: Implement
    saveEvent(event) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (event.eventType === models_1.EventTypes.CONNECTION) {
                yield this.connectionEventEntity.put(event);
            }
            if (event.eventType === models_1.EventTypes.MESSAGE) {
                yield this.messageEventEntities[event.type].put(event);
            }
            // TODO: Implementation of other event types
            return event;
        });
    }
    // @ts-ignore
    updateEvent(event) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () { });
    } // TODO: Implement
    // @ts-ignore
    deleteEvent(id) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () { });
    } // TODO: Implement
}
exports.DynamoDBDataProvider = DynamoDBDataProvider;
