// @botonic/react/lib/cjs/webchat/trigger-button/styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TriggerImage = exports.UnreadMessagesCounter = exports.StyledTriggerButton = void 0;
const tslib_1 = require("tslib");
const styled_components_1 = tslib_1.__importDefault(require("styled-components"));
const constants_1 = require("../../constants");
exports.StyledTriggerButton = styled_components_1.default.div `
  cursor: pointer;
  position: fixed;
  background: ${constants_1.COLORS.SOLID_WHITE};
  border-radius: 50%;
  display: flex;
  align-items: center;
  justify-content: center;
  overflow: hidden;
  width: 65px;
  height: 65px;
  bottom: 20px;
  right: 10px;
  padding: 8px;
`;
exports.UnreadMessagesCounter = styled_components_1.default.div `
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 50%;
  position: absolute;
  right: 20px;
  bottom: 80px;
  width: 24px;
  height: 24px;
  background-color: ${constants_1.COLORS.RED_NOTIFICATIONS};
  color: white;
  z-index: 10;
`;
exports.TriggerImage = styled_components_1.default.img `
  max-width: 100%;
  max-height: 100%;
`;
